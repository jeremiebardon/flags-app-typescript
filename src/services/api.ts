import { createApi, fetchBaseQuery } from "@reduxjs/toolkit/query/react";
import { BorderCountry, Country, Region } from "../types/countries.type";

const filterCountries = (countries: Country[]): Country[] => {
  return countries.sort((a, b) => {
    if (a.name.common < b.name.common) {
      return -1;
    }
    if (a.name.common > b.name.common) {
      return 1;
    }
    return 0;
  });
};

export const api = createApi({
  reducerPath: "api",
  baseQuery: fetchBaseQuery({ baseUrl: "https://restcountries.com/v3.1/" }),
  tagTypes: ["Country"],
  endpoints: (builder) => ({
    getCountries: builder.query<Country[], Region>({
      query: (region) =>
        region && region !== "all" ? `region/${region}` : "all",
      providesTags: (result = [], error, arg) =>
        result
          ? [
              ...result.map(({ cca3 }) => ({ type: "Country" as const, cca3 })),
              "Country",
            ]
          : ["Country"],
      transformResponse: (response: Country[], meta, arg) => {
        return response ? filterCountries(response) : [];
      },
    }),
    getCountryByCode: builder.query<Country[], string>({
      query: (code) => `alpha/${code}`,
      providesTags: (result, error, arg) => [{ type: "Country", id: arg }],
    }),
    getBorderCountries: builder.query<BorderCountry[], string[]>({
      query: (code) => `alpha?codes=${code.join(",")}&fields=name,cca3`,
    }),
  }),
});

export const {
  useLazyGetCountriesQuery,
  useGetCountryByCodeQuery,
  useGetBorderCountriesQuery,
} = api;
