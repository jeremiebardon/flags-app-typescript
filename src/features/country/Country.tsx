import {  Box, Link } from '@chakra-ui/react';
import { skipToken } from '@reduxjs/toolkit/query/react';
import { useParams,  Link as RouterLink } from 'react-router-dom';

import { useGetBorderCountriesQuery, useGetCountryByCodeQuery } from '../../services/api';

import { CountryInfo } from '../../components/CountryInfo';
import { ArrowBackIcon } from '@chakra-ui/icons';

export const Country = () => {
  const { countryName } = useParams();

  const { data: countryInfo } = useGetCountryByCodeQuery(countryName ?? skipToken);
  const { data: borderCountries } = useGetBorderCountriesQuery(countryInfo && countryInfo[0].borders ? countryInfo[0].borders : skipToken);

  return (
    <>
      <Box mb={[6, 12]} mt={[2, 4]}>
        <Link as={RouterLink} colorScheme='teal' size='xs' to='/' variant='menu'>
          <ArrowBackIcon /> Back
        </Link>
      </Box>

      { countryInfo &&  <CountryInfo country={countryInfo[0]} bordersCountries={borderCountries} /> }
    </>
  )
}