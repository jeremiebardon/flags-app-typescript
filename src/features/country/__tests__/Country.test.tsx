import { ApiProvider } from "@reduxjs/toolkit/dist/query/react";
import { render } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import { api } from "../../../services/api";
import { Country } from "../Country";

const wrapperProvider = ({ children }: { children: React.ReactNode }): JSX.Element => {
  return (
    <ApiProvider api={api}>
      <BrowserRouter>
        {children}
      </BrowserRouter>
    </ApiProvider>
  )
}

describe('<CountriesList />', () => {
  it('should render Countries List', () => {
    const wrapper = render(<Country />, {
      wrapper: wrapperProvider
    });

    expect(wrapper).toMatchSnapshot();
  })
});
