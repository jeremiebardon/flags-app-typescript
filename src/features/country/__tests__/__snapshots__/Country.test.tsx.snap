// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CountriesList /> should render Countries List 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="css-1liqolt"
      >
        <a
          class="chakra-link css-0"
          href="/"
        >
          <svg
            class="chakra-icon css-onkibi"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
              fill="currentColor"
            />
          </svg>
           Back
        </a>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="css-1liqolt"
    >
      <a
        class="chakra-link css-0"
        href="/"
      >
        <svg
          class="chakra-icon css-onkibi"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
            fill="currentColor"
          />
        </svg>
         Back
      </a>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
